새 PR 보냅니다

siwonkim0 opened the pull request:
😄

siwonkim0 committed:
feat: Keychain에 session ID 존재 유무에 따라 로그인창, 메인화면을 보여주도록 구현 #2

- KeychainManager에서 저장된 정보가 있는지 확인하는 로직 구현
- SceneDelegate에서 Keychain의 session ID 존재 유무에 따라 각각 다른 ViewController를 rootViewController로 설정
- 메인 화면인 MovieListViewController 파일 생성

siwonkim0 committed:
Add: 스토리보드와 파일에 ViewController, ViewModel 생성

- MovieList, MovieDetail, Search, MyAccount ViewController와 각각에 대응하는 ViewModel 생성
- APIManager 싱글턴으로 변경

siwonkim0 committed:
Chore: 폴더 구조 변경

siwonkim0 committed:
feat: session ID로 사용자 계정 정보 가져오기 구현 #1

- MyAccountViewModel에 getAccountDetails 메서드 구현

siwonkim0 committed:
feat: 영화 평점주기 구현 #1

- APIManager 타입에 performDataTaskWithoutDecoding 메서드로 POST한 뒤 성공, 실패여부만 알 수 있도록 구현함
- MovieDetailViewModel 타입에 rateMovie 메서드 구현

siwonkim0 committed:
feat: 사용자 계정 평점목록 불러오기 구현 #1

- MovieDetailViewModel 타입에 getRatedMovies 메서드 구현
- 기존 MovieListItem 모델을 이용하여 디코딩하기 위해서 rating 프로퍼티를 옵셔널 타입으로 추가함
- URLManager의 url에 "?"가 빠져서 난 에러 수정
- MovieListItem 모델의 originalLanguage 프로퍼티 타입을 OriginalLanguage enum에서 String으로 수정함

siwonkim0 committed:
chore: .

siwonkim0 committed:
Removed file that shouldnt be tracked

siwonkim0 committed:
Removed file that shouldnt be tracked

siwonkim0 committed:
Removed file that shouldnt be tracked

siwonkim0 committed:
feat: 영화 평점 삭제 구현 #1

- APIManager과 MovieDetailViewModel에 deleteRating 메서드 구현
- DELETE의 경우에도 POST처럼 성공, 실패여부만 알면 되어서 응답 데이터를 디코딩하지 않고 에러여부만 확인함

siwonkim0 committed:
Removed file that shouldnt be tracked

siwonkim0 committed:
chore: add .gitignore

siwonkim0 committed:
refactor: APIManager 중복코드 함수 분리 #4

- APIManager에 있던 getToken, createSessionId 메서드 AuthVM으로 이동
- 메서드 네이밍 수정
- APIManager에 urlSession 의존성 주입으로 변경

siwonkim0 committed:
refactor: APIManager의 performDataTask 메서드 재사용해서 코드 간결화 #4

- 네트워크 통신을 해서 데이터를 가져오는 메서드인 performDataTaskWithoutDecoding의 네이밍을 performDataTask로 변결
- performDataTask의 결과인 Data를 디코딩하는 decodeDataAfterDataTask 메서드 생성
- 같은 맥락으로 이미지 데이터의 경우 UIImage로 변경하는 convertToUIImageAfterDataTask 메서드 생성

siwonkim0 committed:
refactor: URLManager 타입 재설계 #4

- 기존 URLManger 네이밍을 MovieURL로 변경
- MovieURL은 현재 프로젝트에 필요한 URL을 담아두는 타입
- 새로 만든 URLManager는 URLComponents를 조합해서 URL을 만드는 타입으로 설계하여 어느 프로젝트에서나 재사용이 가능하도록 리팩토링함

siwonkim0 committed:
chore: 코드 정렬

siwonkim0 committed:
Del: 쓸모없는 코드 삭제

siwonkim0 committed:
test. APIManager Unit Test #4

- get, post, delete 메서드 테스트 코드 작성

siwonkim0 committed:
feat: NetworkChecker 타입 생성하여 네트워크 연결이 없을때 테스트 시도하지 않도록 함

siwonkim0 committed:
test: 네트워크 상태와 무관한 테스트 작성

- MovieFinderFakeTests 테스트 클래스 생성
- APIManager의 getData 메서드 테스트
- 테스트에 사용할 JSON 데이터 생성
- URLSessionStub, URLSessionDataTaskStub 생성하여 APIManager에 의존성 주입하여 테스트 코드에서 사용

siwonkim0 committed:
refactor: getImage 메서드 위치 변경 DetailVM -> ListVM #5

- 의미없는 주석 삭제
- getImage 매서드 테스트를 위해 ListVC에 UIImageView 생성
- MovieFinderUnitTests의 Test Double Movie 이름을 MovieDetailData로 변경

siwonkim0 committed:
Chore: 폴더 구조 변경

siwonkim0 committed:
Chore: Entity 타입들 일관성있는 코딩 컨벤션에 맞게 수정

siwonkim0 committed:
chore: Entity 타입 코딩 컨벤션에 맞게 수정

siwonkim0 committed:
Del: MovieListViewModel에서 getLatest 메서드 삭제

siwonkim0 committed:
refactor: NowPlayingMovieList Entity 삭제

siwonkim0 committed:
refactor: 이미지 처리 캐싱 적용 #6

- APIManager에 있던 이미지 관련 메서드인 getImage 삭제
- UIImageView extension과 ImageCacheManager로 이미지의 url을 확인해 캐시에 있으면 그대로 사용하고 없으면 네트워크 통신을 통해 받아오게 구현
- APIManager, MovieListViewModel에 UIKit import -> Foundation으로 변경
- MovieDetail entity 이름을 좀 더 명확하게 TMDBMovieDetail로 변경

siwonkim0 committed:
chore: 네이밍 수정, 주석 삭제

siwonkim0 committed:
feat: usecase, repository 생성 #5

- viewmodel에 있던 네트워크 통신 코드 repository로 이동
- usecase에서 entity를 model로 변경
- MovieListItem 타입에 convertToModel() 메서드 생성

siwonkim0 committed:
feat: Movie List View 구현 #7

- Movie List View 스토리보드로 서치바, 스크롤뷰, 스택뷰 추가
- MovieListCollectionViewController 생성하여 collectionView를 재사용하기위해 XIB로 구현함
- MovieListCollectionViewController를 MovieListViewController의 child로 설정
- MovieListColletionViewCell XIB로 구현

siwonkim0 committed:
feat: MovieListCollectionViewController의 컬렉션뷰에 데이터 연결 #7, #8

- MovieListViewModel의 usecase MovieListCollectionViewModel로 이동
- MovieListCollectionViewController의 UICollecionViewDataSource 메서드 구현
- MovieListCollectionViewCell의 configure 메서드 구현
- ListItem 모델에 프로퍼티 추가
- MovieListCollectionViewCell의 View 컴포넌트의 content hugging/resistance priority 수정하여 포스터 이미지가 레이블을 가리는 문제 해결

siwonkim0 committed:
refactor: ListItem의 originalLanguage 프로퍼티 타입 변경 #9

- OriginalLanguage enum 생성하여 서버 데이터를 뷰에 맞게 변형하여 보여줌

siwonkim0 committed:
feat: 영화 장르 가져오기 구현 #1, #8, #9

- MovieURL에 genre case 추가
- DefaultMoviesRepository에 getNowPlaying, getGenres 의 결과를 합쳐서 genre ID에 따른 장르정보를 가진 ListItem entity 생성
- MovieListCollectionViewCell의 viewmodel인 MovieListCollectionViewItemViewModel 생성하여 MovieListCollctionViewController의 datasource 메서드에서 주입해줌
- MovieListCollectionViewModel의 프로퍼티로 MovieListCollectionViewItemViewModel을 가지게 함
- MovieListCollectionViewModel에서 usecase에서 받아온 entity를 MovieListCollectionViewItemViewModel로 변경

siwonkim0 committed:
refactor: Data Layer는 DTO, Domain Layer는 Entity를 가지도록 네이밍 변경

- 폴더 구조 변경

siwonkim0 committed:
feat: MovieListURL의 case 개수만큼 MovieListViewController의 컬렉션 뷰 생성 #8

- MovieListURL enum을 MovieURL에서 분리해서 생성하여 MovieListViewController에 들어갈 영화 리스트 관리
- MovieListURL의 case 개수만큼 collectionView를 생성하기 위한 의도임
- MovieListCollectionViewController의 viewModel 생성자 주입하기 위해 이니셜라이저 구현
- MovieListViewModel이 usecase를 다시 가지도록 변경
- MovieListViewModel이 MovieListCollectionViewModel 배열을 프로퍼티로 가짐
- MovieListViewModel의 createCollectionViewModels() 메서드로 usecase를 collectionViewModel만들때 주입해줌
- MovieListViewController에서 viewModel이 가진 collectionViewModel 개수만큼 MovieListCollectionViewController 생성하여 스택뷰에 넣음

siwonkim0 committed:
refactor: MovieListViewController에 viewModel 생성자 주입 #8

- instantiateViewController(identifier:, creator:) 메서드 활용하여 커스텀 이니셜라이저를 구현한 스토리보드로 생성되는 뷰컨트롤러를 생성함

siwonkim0 committed:
del: 쓸모없는 코드, 공백 제거

siwonkim0 committed:
fix: APIManager를 싱글톤 클래스로 구현한 부분 삭제

- 임시로 APIManager를 사용하는 ViewModel들에게 APIManager를 프로퍼티로 가지도록 함(추후 리팩토링할 것)

siwonkim0 committed:
refactor: 비동기 처리 async await로 변경 #1

- 기존에는 여러번의 네트워크 통신 결과를 합쳐서 entity를 만들기 위해 중첩된 클로저를 사용하여 가독성이 좋지 않았음
- async await 사용하여 상수에 네트워크 통신 결과를 담아 사용해서 가독성을 높혔음

siwonkim0 committed:
refactor: async 작업을 concurrent 하게 구현 #1

siwonkim0 committed:
fix: genres 레이블에 글자가 공백없이 나오는 부분 해결 #7

siwonkim0 committed:
del: MovieListItem에서 rating 프로퍼티 삭제

- movieListCollectionViewCell에서 rating 정보를 보여주지 않기로 결정

siwonkim0 committed:
refactor: 컬렉션 뷰 셀이 재사용되면서 잘못된 이미지가 보이는 현상 해결 #6, #7

- UIImageView extension을 UIImageView 상속받은 DownlodableUIImageView 클래스로 변경
- 이미지 다운로드 작업 전에 UIImage에 기본값을 넣어줌
- UICollectionViewCell의 prepareForReuse 메서드에서 셀의 이미지뷰에 기본값을 넣고, dataTask를 cancle하여 재사용 되는 셀의 경우 dataTask가 중복으로 일어나지 않도록 처리

siwonkim0 committed:
chore: 쓸모없는 코드 삭제, 파일 이름 변경

siwonkim0 committed:
refactor: 이미지 초기화 작업 위치 변경 #6

- 기존에 prepareForReuse에 있던 셀의 이미지뷰 초기화 코드 삭제
- DownloadableUIImageView에서 이미지뷰 초기화 코드를 이미지가 캐시에 없는 경우 네트워크 통신을 하기 전으로 위치 변경함

siwonkim0 committed:
refactor: Kingfisher 추가하여 디스크 캐싱으로 앱 실행시 이미지 로딩 속도 개선 #6

siwonkim0 committed:
refactor: MovieListCollectionView DataSource를 UICollectionViewDiffableDataSource로 변경 #7

siwonkim0 committed:
feat: Coordinator 패턴 적용 #3

- 로그인 유무에 따라 AuthVC나 MovieListVC를 root뷰로 설정
- Coordinator들을 관리하는 AppCoordinator, AuthVC 관리하는 AuthCoordinator, MovieListVC 관리하는 MovieListCoordinator 생성

siwonkim0 committed:
chore: 네이밍 변경, 폴더 생성

- Authentication -> Auth로 변경
- Presentation 폴더 안에 Coordinator 폴더 생성하여 관련 파일들 이동

siwonkim0 committed:
refactor: child coordinator의 역할이 끝났을 때 parent coordinator에서 제거

- AppCoordinator가 childCoordinators 배열을 가짐
- 화면 전환시 해당 coordinator가 AppCoordinator의 childCoordinators에 append됨
- 해당 vc의 viewDidDisappar 시점에 AppCoordinator에게 알려서 childCoordinators 배열에서 제거
- AuthVC의 delegate 프로퍼티 coordinator로 네이밍 변경

siwonkim0 committed:
chore: 코딩 컨벤션 통일, 쓸모없는 코드 삭제

siwonkim0 committed:
chore: 폴더 위치 변경

siwonkim0 committed:
refactor: 데이터 바인딩 rx로 변경 #10

- APIManager 기존 async await 코드 rxSwift로 변경
- MovieListCollectionViewModel이 itemViewModels를 가지고 있지 않고 받아온 movieList 데이터를 바
로 MovieListViewController의 CollectionView SnapShot에 적용하도록 수정
- AuthViewModel에 transform(), AuthViewController configureBind() 생성하여 뷰컨트롤러와 뷰모델
간 바인딩
- 뷰컨 -> 뷰모델 Input으로 버튼 탭 이벤트를 받아서 뷰모델 -> 뷰컨으로 output 전달
- SceneDelegate의 프로퍼티로 AppCoordinator을 가지게 하고, AuthCoordinator가 가진 parentCoordinator 에 weak 키워드 추가하여 순환참조 방지함

siwonkim0 committed:
fix: AuthViewModel의 Input 바인딩 오류 수정

- input으로 들어오는 didTapAuthDone 이벤트를 받으면 뷰모델에서 sessionID를 만들고, 키체인에 저장한 후에 output으로 완료 이벤트를 전달하는 로직
- 기존에는 CombineLstest로 tap이벤트와 sessionID 생성 이벤트를 합쳤으나, tap 이벤트를 받으면 sessionID를 생성해서 sessionID를 이벤트로 전달하는 방향으로 수정함

siwonkim0 committed:
feat: ViewModelType 프로토콜 생성하여 AuthViewModel에 채택

- AuthViewModel의 transform 메서드 위치 변경

siwonkim0 committed:
feat: Unit Test를 위한 RxTest, RxNimble 설치

siwonkim0 committed:
refactor: MovieListCollectionView를 viewModel과 바인딩 방법 변경 #10

- viewWillAppear 시점에 api call을 하도록 viewWillAppear observable을 viewModel이 받아 api call 후에 뷰컨에 output 데이터를 전달하여 뷰에 데이터를 채워줌

siwonkim0 committed:
refactor: Movie List View의 CollectionView를 구현하는 방식 변경 #7

- 기존에는 MovieListViewController가 child ViewController 여러개를 가지고, 한 뷰컨당 컬렉션뷰 하나를 가지도록 구현
- UICollectionViewCompositionalLayout을 이용하여 MovieListViewController는 UICollectionView 하나만 가지고, 컬렉션뷰 내부에서 여러 section을 구현함
- MovieListView에 스크롤뷰, 스택뷰 삭제후 컬렉션뷰 추가

siwonkim0 committed:
feat: MovieListURL 개수만큼 section을 형성하여 NSDiffableDataSourceSnapshot에 적용 #7

- Section 클래스 생성하여 movies 배열을 가지고 있도록 함
- MovieListViewModel의 Output을 [MovieListCollectionViewModel] -> [Section]으로 변경
- Observable.zip을 이용해 Observable<Section> 시퀀스를 하나의 Observable<[Section]>으로 변환
- MovieListItemViewModel에 id, section 프로퍼티 추가하여 같은 영화가 여러개의 영화목록에 중복으로 존재할 수 있도록 구현

siwonkim0 committed:
chore: private, final 키워드 추가

siwonkim0 committed:
style: 함수 줄바꿈 변경

siwonkim0 committed:
feat: UICollectionView에 supplementaryView 구현하여 section title 설정 #7

- MovieListHeaderView 생성
- MovieListHeaderView를 collectionView에register
- datasource supplementaryProvider 프로퍼티 설정하여 데이터 뿌려줌
- UICollectionViewCompositionalLayout의 section에 boundarySupplementaryItems 추가

siwonkim0 committed:
refactor: Coordinator Pattern 추가 적용 #3

- MyAccountCoordinator, SearchCoordinator 생성
- 기존에는 모든 coordinator에 window참조를 전달하여 각자가 가진 뷰컨을 생성하여 window의 rootviewcontroller로 설정했음
- 그러나 AppCoordinator만 window를 가지고 있고, child coordinator들은 뷰컨을 생성하여 AppCoordinator에 전달하면 AppCoordinator가 대신 rootView로 설정하도록 변경
- Coordinator 프로토콜에 childCoordinators 프로퍼티 추가

siwonkim0 committed:
feat: MovieListController에 NavigationController 연결 #3

- MovieListCoordinator가 화면전환시 navigationController에 MovieListController를 add해서 화면전환
- NavigationBat Title 설정

siwonkim0 committed:
feat: AppCoordinator에 UITabBarController를 생성하는 tabBarController() 메서드 구현 #3

- tabBarController에 ListVC 추가

siwonkim0 committed:
feat: UITabBarController에 ViewControllers 추가 #3

- AppCoordinator에서 UITabBarController에 SearchViewController, MyAccountViewController 추가 연결
- SearchViewController, MyAccountViewController에 coordinator 프로퍼티 추가하고 프로토 coordinator에 각각 채택시킴
- SearchCoordinator, MyAccountCoordinator에 각각 UINavigationController를 반환하는 setViewController() 메서드 생성

siwonkim0 committed:
feat: Tab Bar Item image SF Symbols에서 추가 #11

siwonkim0 committed:
test: MovieListViewModelTest 생성

- FakeDefaultMoviesUseCase 생성하여 viewModel에 주입

siwonkim0 committed:
chore: 함수 네이밍 수정, 쓸모없는 코드 삭제

siwonkim0 committed:
chore: 쓸모없는 코드 삭제

siwonkim0 committed:
refactor: AuthViewController에 viewModel 주입하도록 수정 #3

- viewModel에 repository 주입
- repository에 apiManager 주입
- AuthVC에 이니셜라이저 생성하여 viewModel 주입
- AuthCoordinator에 AuthVC 인스턴스 생성 부분 새로운 이니셜라이저로 수정

siwonkim0 committed:
test: AuthViewModelTest 생성

- AuthRepository에 MovieAuthRepository 프로토콜 채택
- AuthViewModel 테스트를 위해 FakeAuthRepository 생성

siwonkim0 committed:
refactor`: repository interface 파일 분리

siwonkim0 committed:
refactor: KeyChain에 session id 저장하는 메서드 repository로 위치 변경

siwonkim0 committed:
refactor: viewWillAppear 이벤트를 한번만 받도록 take 사용

- 다른 tab으로 이동했다가 movieListVC로 돌아오면 viewMillAppear가 다시 불려서 collectionView가 reload되는 현상 발생하여 최초에 한번만 이벤트를 받도록 수정

siwonkim0 committed:
feat: MovieListVC에 refreshController 적용 #7

- Input으로 refresh event 받아서 viewModel에서 데이터 fetch 후 Output으로 전달하여 collectionview reload

siwonkim0 committed:
feat: MovieDetailGlobalHeaderView XIB 생성 #12

siwonkim0 committed:
feat: MovieListCollectionView의 셀 클릭시 MovieDetailViewController로 화면전환 #3, #7

- 셀 클릭 이벤트가 발생시 list view의 coordinator에게 화면전환 요청
- AppCoordinator에 detailVC 인스턴스 생성 로직 구현

siwonkim0 committed:
feat: MovieDetailViewController에 viewModel 주입, repository에서 영화 상세정보 가져오는 로직 구현 #12

- DefaultMoviesRepository의 getMovieDetail() 메서드로 DTo -> entity 데이터 매핑하여 리턴
- OMDBMovieDetailDTO에 convertToEntity() 메서드 구현
- detailView에 필요한 여러 entity 타입들 구현
- MovieDetailViewModel 이니셜라이저 생성하여 movieID, useCase 주입
- MovieDetailViewController 이니셜라이저 생성하여 viewModel 주입
- MovieListViewController didSelectedItem에서 selectedModel 구하는 로직 수정

siwonkim0 committed:
feat: MovieDetailViewController collectionView에 review 정보 가져오기 구현 #12

- DefaultMoviesRepository에서 MovieDetailReview 배열을 반환하는 getMovieDetailReviews() 구현
- MovieDetailViewModel과 MovieDetailViewController을 Input-Output으로 바인딩
- MovieDetailViewController에 collectionView 구현
- 각 section마다 다른 셀과 다른 데이터를 보여주도록 DiffableDataSource의 SectionIdentofier, ItemIdentifier을 enum으로 구현
- MovieDetailGlobalHeaderView XIB 생성
- MovieDetailCommentsCollectionViewCell XIB 생성

siwonkim0 committed:
feat: collectionView Section Header 생성하여 등록 #12

- collectionView에 header 등록, 관련 layout 설정, section마다 다른 제목 구현
- SectionBackgroundDecorationView 생성하여 section background 설정

siwonkim0 committed:
refactor: UICollectionView extension 구현

-  UICollectionViewCell, UICollectionReusableView register, dequeue 메서드 구현하여 코드 간결화

siwonkim0 committed:
refactor: MovieDetailViewController Storyboard view 구조 변경 #12

- MovieDetailGlobalHeaderView XIB 삭제
- Storyboard에 스크롤뷰-스택뷰안에 영화 기본정보를 담은 스택뷰 생성
- 영화 기본정보 스택뷰 아래에는 컬렉션뷰 생성하여 comments section 보여줌

siwonkim0 committed:
feat: MovieDetailViewController 기본 영화정보 불러오기 구현 #12

- plot summary cell XIB 생성하여 컬렉션뷰에 register
- 기존에는 메서드 안에서 생성해주던 snapshot 프로퍼티를 MovieDetailViewController가 가지고있도록 하여 여러번 apply 가능하도록 수정
- MovieDetailViewController에서 output인 basicInfoObservable을 subscribe하여 뷰 업데이트 구현

siwonkim0 committed:
feat: UIImage를 UIImageView에 맞게 리사이징, UIImageView cornerRadius 설정 #12

- 이미지 캐시전에 서버에서 받아온 이미지를 resize해서 캐시함

siwonkim0 committed:
feat: MovieListCollectionViewCell의 UIImageView에도 cornerRadius 설정 #7

- MovieListCollectionViewCell 오토레이아웃 수정: UIImageView의 비율을 image의 비율에 맞게 설정하고 UIImageView를 stackView에 넣어서 Alignment를 center로 맞춤
- MovieDetailViewController에 configureImageView 메서드 생성: Kingfisher로 placeHolder, transition 설정

siwonkim0 committed:
chore: 주석 제거

siwonkim0 committed:
fix: MovieDetailVC에서 viewWillAppear이벤트가 있을때마다 뷰를 업데이트하는 부분 수정 #12

- observable에 take(1) operator을 사용하여 최초 한번만 뷰 업데이트하도록 구현

siwonkim0 committed:
refactor: MovieDetailVC collectionView section마다 다른 layout 적용 #12

- createLayout() 메서드 수정하여 section별로 다른 layout 적용

siwonkim0 committed:
refactor: supplementaryHeaderItem 메서드 분리 #12

- section에 각각 contentInsets를 줘서 leading, trailing 부분 간격 생성

siwonkim0 committed:
refactor: collectionView contentSize에 맞는 height 적용 #12

siwonkim0 committed:
feat: comment 더보기 기능 구현 #12

- MovieDetailReview를 id로 비교하도록 hash함수 구현
- MovieDetailViewController didSelectedItem() 메서드 구현하여 셀이 터치되면 commentLabel의 길이가 변하도록 구현(리팩토링 필요)

siwonkim0 committed:
chore: 줄바꿈 수정

siwonkim0 committed:
refactor: comment 더보기 기능 수정 #12

- comment 셀 터치시, layoutIfNeeded 대신 collectionViewLayout invalidLayout 사용하여 새로 업데이트된 레이아웃 적용함

siwonkim0 committed:
refactor: comment의 글자 수가 300개 이하인경우 모든 글자를 보여주도록 수정 #12

- 300개 이하인경우 index out of range 오류 해결

siwonkim0 committed:
refactor: 컬렉션 뷰 셀의 높이가 변경되면 컬렉션뷰 전체 높이도 변경 #12

- comment 셀 터치시, 그 시점의 컬렉션뷰의 contentSize.height를 계산해서 적용함

siwonkim0 committed:
feat: Cosmos 라이브러리 pod install #13

- 평점주기 기능 구현을 위한 Cosmos 라이브러리 등록

siwonkim0 committed:
feat: MovieDetailVC 서버에서 받아온 평점 표시 #12

- 10점 만점의 평점을 5점 만점으로 환산하여 반올림한 평점을 별 5개에 표시
- 별 반개 평점도 줄 수 있도록 구현

siwonkim0 committed:
feat: comments 부분에도 서버에서 받아온 평점 표시 #12

- MovieDetailVC의 영화 평점은 서버에서 받아온 값은 레이블로 표시하고, ratingView는 사용자가 직접 평점줄 수 있도록 변경

siwonkim0 committed:
design: 평점 버튼 테두리 색 변경

siwonkim0 committed:
feat: 평점 버튼 터치시 서버에 업데이트 #12

- RateRespondDTO 타입 생성
- MovieDetailViewModel의 Input에 tapRatingButton Observable 생성, Output에 ratingDriver Driver 생성
- MovieDetailViewController에 ratingRelay 생성하여 평점 버튼 탭 이벤트 발생시 이벤트 전달함
- input tapRatingButton 이벤트 발생시 서버 업데이트
- output ratingDriver 이벤트 발생시 뷰 업데이트

siwonkim0 committed:
feat: 저장된 평점을 불러오기 구현 #12

- MovieDetailViewController의 viewWillAppear시점에 사용자가 평점을 준 영화목록을 불러와서 해당 영화가 목록에 있으면 평점을 불러오도록 구현
- RatedMovieDTO 생성

siwonkim0 committed:
feat: account id 키체인에 저장 구현 #13

- AuthViewController의 viewWillAppear 시점에 키체인에 저장되어있는 session ID를 이용하여 account ID를 가져온 후 키체인에 저장
- AccountRepository 생성하여 AuthViewModel이AuthRepository, AccountRepository 둘 다를 가지도록 구현

siwonkim0 committed:
refactor: 키체인에서 가져온 account id로 평점목록을 가져오도록 수정 #12

siwonkim0 committed:
refactor: AuthViewModel에 있던 session id 저장 부분을 AuthRepository로 이동 #14

siwonkim0 committed:
refactor: AuthViewModel의 비즈니스 로직을 AuthUseCase 생성하여 분리 #13

siwonkim0 committed:
chore: 삭제된 파일 다시 생성

siwonkim0 committed:
chore: ds_store 제거

siwonkim0 committed:
feat: SearchViewController에 UISearchController 생성 #15

- 검색 결과를 보여줄 tableView를 searchResultsController로 연결
- snapkit 라이브러리 사용

siwonkim0 committed:
feat: SearchViewController에 collectionView 생성 #15

- RecentSearchCollectionViewCell 생성
- SearchViewController에 collectionView configure

siwonkim0 committed:
chore: 상속하지 않는 class에 final 키워드 추가

siwonkim0 committed:
chore: 순환 참조 방지 위해 weak self 처리

siwonkim0 committed:
chore: 코드 위치 수정

siwonkim0 committed:
refactor: 스토리보드 제거

siwonkim0 committed:
refactor: MovieDetailViewController 스토리보드 삭제

- SnapKit 사용하여 코드로 대체

siwonkim0 committed:
refactor: Diffable DataSource reconfigureItems 활용해 item 업데이트 #12

- collectionview item identifier을 Identifiable 프로토콜을 채택한 Struct의 ID 타입으로 변경
- review 더보기 버튼 터치시 전체 리뷰가 보여지도록 구현
- 기존에는 cell 단에서 보여지는 글자 수를 조절했다면, 이번엔 데이터 자체를 다르게 바꿔서 diffable datasource가 변경사항을 감지할 수 있도록 변경

siwonkim0 committed:
chore: 변수명 rename

- comment, review 혼용되는 부분 review로 통일

siwonkim0 committed:
chore: 오류 해결

siwonkim0 committed:
fix: collectionview의 comment 부분 높이가 커지면 collectionview가 잘리는 현상 해결 #12

- 셀 터치 이벤트 발생시 updateConstraints로 collectionview의 높이를 contentview의 높이와 동일하게 업데이트

siwonkim0 committed:
fix: collectionview 잘리는 현상 버그 수정 #12

- viewDidAppear에서 content가 전부 로드되고 나서 contentView의 높이를 계산하여 collectionview의 높이로 설정해줌

siwonkim0 committed:
refactor: 네트워크 계층 추상화 #1

- URLRequest를 생성하는데 필요한 요소들을 NetworkRequest 프로토콜로 추상화
- 각 URLRequest 생성하는 타입은 NetworkRequest를 채택
- AuthUseCase의 네트워크 관련 로직을 AuthRepository로 분리

siwonkim0 committed:
refactor: 새로운 네트워크 코드 detail에 적용

- 영화 평점 등록하기, 가져오기 로직 account repository로 이동
- account repository를 detail usecase에 주입

siwonkim0 committed:
refactor: 새로운 네트워크 코드 image에 적용

siwonkim0 committed:
refactor: 새로운 네트워크 코드 list에 적용

siwonkim0 committed:
chore: 네이밍 수정

siwonkim0 committed:
refactor: listVC에 영화 목록 title 설정 #8

siwonkim0 committed:
refactor: 모든 평점 목록 페이지를 가져오도록 수정 #12

- 평점 준 영화의 개수가 많아지면서 평점 목록의 페이지수가 늘어나게 되었음
- 기존에는 첫페이지만 가져와서 나중에 평점을 준 영화는 해당 목록에 없게 되어 평점을 불러올 수 없었다
- 따라서 RatedMovieListDTO의 totalPages 프로퍼티를 활용하여 totalPages 수만큼 네트워크 요청을 하여 모든 평점 목록을 가져와서 그 안에 해당 영화 ID가 존재하는지 찾도록 구현

siwonkim0 committed:
fix: ratingView의 기본값 설정 #12

- ratingView 의 기본값 0으로 설정
- 평점 반개 단위도 줄 수 있도록 설정

siwonkim0 committed:
deaign: 앱 전체 배경색을 흰색으로 변경

- NSLayoutConstraint 코드를 Snapkit으로 변경
- AuthController의 Storyboard에 시작화면 추가
- CodmosView 기본 세팅 변경
- listView의 collectionviewLayout group 크기 변경
- font bold로 변경

siwonkim0 committed:
refactor: Observable 스트림 공유하도록 변경 #12

- detail viewModel에서 넘겨준 Observable을 viewController에서 UI 컴포넌트들에 bind
- plot summary를 collectionview에서 보여주던 것을 collectionview 밖에서 보여주도록 변경

siwonkim0 committed:
feat: SearchViewController의 collectionview에 검색 결과가 표시되도록 구현 #15

- UICollectionViewListCell 을 이용하여 list collectionview 구현
- DefaultMoviesRepository에서 키워드 기반 검색결과를 반환하는 getSearchMovieList() 메서드 구현

siwonkim0 committed:
chore: 주석 삭제

siwonkim0 committed:
refactor: search시 error가 나면 Observable이 dispose되는 현상 해결 #15

- 검색어가 빈 문자열(””)일때 네트워크 요청시 error가 나서 filter 사용하여 빈문자열의 경우 검색요청이 일어나지 않도록 수정
- ObservableType+extension 을 생성하여 error가 났을 때 무시하여 스트림이 끊어지지 않도록함
- search 결과를 collectionview에 표시하는데 snapshot apply시 main thread에서 일어나도록 수정하여 화면이 멈추는 현상 해결

siwonkim0 committed:
feat: searchBar의 cancel 버튼 터치시 검색 결과가 사라지도록 구현 #15

siwonkim0 committed:
refactor: listCell에서 url생성해주는 부분 cellViewModel이 하도록 분리 #8

siwonkim0 committed:
chore: MovieListItemViewModel -> MovieListCellViewModel로 이름 변경

siwonkim0 committed:
refactor: Image downsample, SearchViewController에 viewModel 주입, searchCollectionViewCell 커스텀 셀로 변경 #15

- SearchCollectionViewCell nib 파일 생성하여 searchCollectionView에 register
- UIImage extension으로 downSample 메서드 구현하여 서버에서 받아온 큰 사이즈의 이미지를 썸네일 형태로 작게 가공함

siwonkim0 committed:
refactor: collectionview layout 깨지는 현상 해결, 이미지 캐시 #15

- UICollectionViewCompositionalLayout list 대신 UICollectionViewFlowLayout 사용
- KingFisher의 DownsamplingImageProcessor 사용하여 이미지 다운샘플링
- 이미지 캐시

siwonkim0 committed:
feat: 이미지 downsampling list, detail의 이미지에도 적용 #6

siwonkim0 committed:
feat: 검색 결과 터치하면 상세화면 보여주도록 추가 #15

- poster 정보가 없는 경우 검색 결과에서 제외하도록 수정

siwonkim0 committed:
refactor: Auth 화면 디자인 변경 #14

siwonkim0 committed:
chore: 주석 삭제

siwonkim0 committed:
refactor: 로그인 완료후 앱으로 진입시 메인화면으로 전환되도록 구현 #14

- sceneWillEnterForeground 이벤트를 NotificationCenter로 구독하고 이벤트가 발생하면 KeyChain에 sessionId 저장여부를 확인하여 로그인 완료시 메인화면으로 화면전환하도록 구현

siwonkim0 committed:
chore: 쓸모없는 코드 삭제

siwonkim0 committed:
design: detail view 평점 글자크기 조절 #12

siwonkim0 committed:
feat: detail view 평점 등록시 alert 띄우도록 구현 #12

- 사용자에게  평점이 등록됐다는 사실을 알리기 위해 alert 추가

siwonkim0 committed:
feat: NSMutableAttributedString 으로 커스텀 폰트 적용

- detail 화면의 리뷰길이가 길면 preview에 ...more 이 추가되도록 하기위해 NSMutableAttributedString 사용
- NSMutableAttributedString extension 글자마다 다른 폰트를 적용할 수 있도록 구현

siwonkim0 committed:
refactor: ratedMovieListDTO -> movieListDTO로 대체

siwonkim0 committed:
design: 줄바꿈 수정

siwonkim0 committed:
refactor: Repository에 있는 비즈니스로직 usecase로 이동

siwonkim0 committed:
refactor: Repository에 있는 비즈니스로직 usecase로 이동

- MovieDetailReview entity를 repository가 반환하고 있던 부분을 usecase로 이동

siwonkim0 committed:
refactor: Repository에 있는 비즈니스로직 usecase로 이동

- MovieListItem entity를 repository가 반환하고 있던 부분을 usecase로 이동
- SearchCellViewModel을 usecase가 반환하고 있던 부분을 viewModel로 이동

siwonkim0 committed:
refactor: Repository에 있는 비즈니스로직 usecase로 이동

- MovieListItem entity를 repository가 반환하고 있던 부분을 usecase로 이동
- view에 관련된 모델인 Section을 usecase가 반환하고 있던 부분을 viewModel로 이동

siwonkim0 committed:
refactor: AccountUseCase 생성하여 로직 분리

yim2627 commented on MovieFinder/Application/AppDelegate.swift:
```diff
@@ -14,6 +14,12 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
+        if #available(iOS 15.0, *) {
+            let appearance = UITabBarAppearance()
+            appearance.configureWithOpaqueBackground()
+            UITabBar.appearance().standardAppearance = appearance
+            UITabBar.appearance().scrollEdgeAppearance = appearance
+        }
```
왜 appearance를 변경해주나요?

siwonkim0 replied:
[공식문서](https://developer.apple.com/documentation/technotes/tn3106-customizing-uinavigationbar-appearance)에 따르면, iOS 15부터 네비게이션바의 기본값이 투명으로 바뀌어서 검은색으로 보여지는 이슈가 있었습니다. 따라서 불투명으로 바꿔주었습니다. 

[UINavigationBar](https://developer.apple.com/documentation/uikit/uinavigationbar) in iOS 15 introduces changes to its appearance settings. It extends the usage of its [scrollEdgeAppearance](https://developer.apple.com/documentation/uikit/uinavigationbar/3198027-scrolledgeappearance), which by default produces a transparent background, to all navigation bar styles.

yim2627 commented on MovieFinder/Application/SceneDelegate.swift:
```diff
@@ -10,13 +10,18 @@ import UIKit
 class SceneDelegate: UIResponder, UIWindowSceneDelegate {
```
상속을 하지않는 다른 클래스들은 전부 `final` 키워드를 붙여주셨는데, AppDelegate와 SceneDelegate에는 안붙여주신 이유가 있나요?

yim2627 commented on MovieFinder/Application/SceneDelegate.swift:
```diff
@@ -10,13 +10,18 @@ import UIKit
         // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
-        guard let _ = (scene as? UIWindowScene) else { return }
+        guard let windowScene = (scene as? UIWindowScene) else { return }
+//        KeychainManager.shared.deleteExistingSession()
+//        KeychainManager.shared.deleteAccountId()
```
필요없는 주석, 코드들은 전부 지워주시는 게 좋을 것 같아요.

siwonkim0 replied:
이건 로그인 테스트할때 필요한거라 주석처리해놓았습니다

yim2627 commented on MovieFinder/Application/SceneDelegate.swift:
```diff
@@ -10,13 +10,18 @@ import UIKit
+        coordinator?.start()
     }
 
     func sceneDidDisconnect(_ scene: UIScene) {
```
사용하지 않는 메서드도 마찬가지로 지워주시는게 좋을 것 같아요.

yim2627 commented on MovieFinder/Application/AppDelegate.swift:
```diff
@@ -14,6 +14,12 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
```
이 파일엔 의미 없는 줄바꿈이 많은 것 같아요.

yim2627 commented on MovieFinder/Data/DTO/Account/AccountDetailDTO.swift:
```diff
@@ -0,0 +1,48 @@
+    let iso3166_1: String
+    let name: String
+    let includeAdult: Bool
+    let username: String
```
Camel Case

```suggestion
    let userName: String
```

yim2627 replied:
CodingKey 바꾸는 것도 까먹지 마시고요.

yim2627 commented on MovieFinder/Data/DTO/Account/AccountDetailDTO.swift:
```diff
@@ -0,0 +1,48 @@
+    let iso639_1: String
+    let iso3166_1: String
+    let name: String
+    let includeAdult: Bool
```
```suggestion
    let isIncludeAdult: Bool
```

yim2627 commented on MovieFinder/Data/DTO/Account/AccountDetailDTO.swift:
```diff
@@ -0,0 +1,48 @@
+import Foundation
+
+// MARK: - AccountDetailDTO
+struct AccountDetailDTO: Codable {
```
Codable이 붙었다는건 인코딩, 디코딩을 다 한다는 것인데 그 용도가 맞나요?

siwonkim0 replied:
디코딩만 하는 모델은 Decodable을 채택하도록 변경하였습니다

yim2627 commented on MovieFinder/Data/DTO/Account/RateRespondDTO.swift:
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+struct RateRespondDTO: Decodable {
+    let success: Bool
```
```suggestion
    let isSuccess: Bool
```

yim2627 commented on MovieFinder/Data/DTO/Authentication/RequestToken.swift:
```diff
@@ -0,0 +1,15 @@
+//  Created by Siwon Kim on 2022/05/18.
+//
+
+import Foundation
```
줄바꿈 통일성

siwonkim0 replied:
Thank you

yim2627 commented on MovieFinder/Data/DTO/Authentication/Session.swift:
```diff
@@ -0,0 +1,24 @@
+
+// MARK: - Session
+struct Session: Codable {
+    let success: Bool
+    let failure: Bool?
```
```suggestion
    let isSuccess: Bool
    let isFailure: Bool?
```

yim2627 commented on MovieFinder/Data/DTO/Authentication/Token.swift:
```diff
@@ -0,0 +1,21 @@
+
+// MARK: - Token
+struct Token: Codable {
+    let success: Bool
```
```suggestion
    let isSuccess: Bool
```

yim2627 commented on MovieFinder/Data/DTO/MovieDetail/ReviewListDTO.swift:
```diff
@@ -0,0 +1,61 @@
+// MARK: - AuthorDetailsDTO
+struct AuthorDetailsDTO: Codable {
+    let name: String
+    let username: String
```
```suggestion
    let userName: String
```

yim2627 commented on MovieFinder/Data/DTO/MovieDetail/TMDBMovieDetailDTO.swift:
```diff
@@ -0,0 +1,121 @@
+    let runtime: Int
+    let spokenLanguages: [SpokenLanguageDTO]
+    let status: String
+    let tagline: String
```
```suggestion
    let tagLine: String
```

yim2627 commented on MovieFinder/Data/DTO/MovieDetail/VideoListDTO.swift:
```diff
@@ -0,0 +1,41 @@
+    let site: String
+    let size: Int
+    let type: String
+    let official: Bool
```
```suggestion
    let isOfficial: Bool
```

yim2627 commented on MovieFinder/Data/ImageCacheManager.swift:
```diff
@@ -0,0 +1,13 @@
+final class ImageCacheManager {
+    static let shared = NSCache<NSString, UIImage>()
+    private init() {}
+}
```
KingFisher를 쓰는 것 같으니 이 타입은 필요가 없겠네요. 쓰는 곳이 없는 것 같아요.

yim2627 commented on MovieFinder/Data/KeychainManager.swift:
```diff
@@ -0,0 +1,131 @@
+    }
+    
+    @discardableResult
+    private func read(service: String, account: String) throws -> Data? {
+        let query: [String: AnyObject] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrService as String: service as AnyObject,
+            kSecAttrAccount as String: account as AnyObject,
+            kSecReturnData as String: kCFBooleanTrue,
+            kSecMatchLimit as String: kSecMatchLimitOne
+        ]
+        var result: AnyObject?
+        let status = SecItemCopyMatching(query as CFDictionary, &result)
+        
+        guard status != errSecItemNotFound else {
+            throw KeychainError.itemNotFound
+        }
+        guard status == errSecSuccess else {
+            throw KeychainError.unexpectedStatus(status)
+        }
+        
+        return result as? Data
+    }
```
해당 메서드에서 데이터를 불러올 때 발생할 수 있는 에러(Data 오류)이니 옵셔널로 반환하지않고, 내부에서 throw하는 것이 좋아보여요 .
물론 외부에서 에러를 처리하는 것이지만, 여기서 에러를 지정하여 던지면 더 가독성이 좋아질 것 같아요.

siwonkim0 replied:
한곳에 에러처리를 모아놓으니 보기 편해졌어요~!

yim2627 commented on MovieFinder/Data/Network/Error/URLSessionError.swift:
```diff
@@ -8,18 +8,24 @@
         case .requestFailed(description: let description):
             return description
         case .responseFailed(code: let code):
             return "statusCode: \(code)"
```
굳이 연관값 파라미터는 쓰지않아도 괜찮아보일 것 같네요.

yim2627 commented on MovieFinder/Data/Network/HttpMethod.swift:
```diff
@@ -0,0 +1,31 @@
+//
+//  URLRequest+extension.swift
```
이름이 다른게 많아요.

yim2627 commented on MovieFinder/Data/Network/Service/AuthenticationService.swift:
```diff
@@ -0,0 +1,106 @@
+//                    DispatchQueue.main.async {
+//                        UIApplication.shared.open(url, options: [:], completionHandler: nil)
+//                    }
+//                }
```
필요없는 코드

yim2627 commented on MovieFinder/Util/DownloadableUIImageView.swift:
```diff
@@ -0,0 +1,50 @@
+
+import UIKit
+
+class DownloadableUIImageView: UIImageView {
```
다른 곳엔 다 붙어있는데 이 타입은 final이 안붙어있네요

yim2627 commented on MovieFinder/Data/Network/URLSessionManager.swift:
```diff
@@ -0,0 +1,50 @@
+import RxSwift
+
+final class URLSessionManager {
+    var urlSession: URLSessionProtocol
```
다른 곳에서 접근하거나 값을 변경해주나요? 보기엔 이니셜라이저로 주입해주는 것밖에 보이지않네요.

yim2627 commented on MovieFinder/Data/Repository/AccountRepository.swift:
```diff
@@ -0,0 +1,112 @@
+import RxSwift
+
+final class AccountRepository: MovieAccountRepository {
+    let urlSessionManager: URLSessionManager
```
다른 곳에서 접근하지않는 것 같네요.

yim2627 commented on MovieFinder/Data/Repository/AuthRepository.swift:
```diff
@@ -0,0 +1,67 @@
+
+final class AuthRepository: MovieAuthRepository {
+    let urlSessionManager: URLSessionManager
+    var token: String?
```
다른 곳에서 접근하지않는 것 같네요.

yim2627 commented on MovieFinder/Data/Repository/AuthRepository.swift:
```diff
@@ -0,0 +1,67 @@
+                    return
+                }
+                self.saveSessionIDToKeychain(dataSessionID)
+            }
```
문제가 생길 것 같은 코드네요.

siwonkim0 replied:
weak self 붙여주었습니다~!

yim2627 commented on MovieFinder/Data/Repository/DefaultMoviesRepository.swift:
```diff
@@ -0,0 +1,70 @@
+    func getOmdbMovieDetail(with id: Int) -> Observable<OMDBMovieDetailDTO> {
+        let tmdbRequest = DetailMovieInfoRequest(urlPath: "movie/\(id)?")
+        return urlSessionManager.performDataTask(with: tmdbRequest)
+            .withUnretained(self)
+            .flatMap { (self, detail) -> Observable<OMDBMovieDetailDTO> in
+                let id = detail.imdbID ?? ""
+                let omdbRequest = DetailOmdbMovieInfoRequest(
+                    urlPath: "?\(id)",
+                    queryParameters: [
+                        "i": "\(id)",
+                        "apikey": ApiKey.omdb.description
+                    ])
+                return self.urlSessionManager.performDataTask(with: omdbRequest)
```
개인적인 의견이지만 withUnretained를 사용하여 얻은 인스턴스 네이밍을 self로 해버려서 이게 unretained 메서드를 거친 self인지 아닌지가 엄청 헷갈리네요.

yim2627 commented on MovieFinder/Domain/Entity/MovieDetailReview.swift:
```diff
@@ -0,0 +1,19 @@
+
+struct MovieDetailReview: Identifiable, Hashable {
+    let id: UUID = UUID()
+    let username: String
```
```suggestion
    let userName: String
```

yim2627 commented on MovieFinder/Domain/Entity/MovieDetailReview.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+struct MovieDetailReview: Identifiable, Hashable {
```
`Identifiable`을 채택받은 이유는 무엇인가요?

siwonkim0 replied:
Diffable Datasource의 item identifier로 MovieDetailReview.ID 타입을 지정해주기 위해서입니다. 

영화 리뷰에 더보기 기능을 구현하기 위해서 셀을 터치할때마다 item의 내용이 변경되어야 했습니다.
기존 셀의 내용을 업데이트할 때 새로운 셀로 대체하지 않고 해당 셀의 내용만 변경하려면 [reconfigureItems(_:)](https://developer.apple.com/documentation/uikit/nsdiffabledatasourcesnapshot/3804468-reconfigureitems) 을 통해 가능합니다. 불필요하게 셀을 삭제하거나 새로 삽입하지 않아 성능을 높일 수 있습니다.

하지만, 이때 item identifier가 struct라면 이 메서드를 사용할 수 없기 때문에 MovieDetailReview에 Identifiable을 채택하여 associatedType인 MovieDetailReview.ID 을 Diffable Datasource의 item identifier로 지정해주었습니다.

<img width="779" alt="스크린샷 2022-08-30 오후 6 47 00" src="https://user-images.githubusercontent.com/60725934/187405897-645e01a6-83a6-4ae2-a7f1-03e11a5ba1ee.png">

[관련 문서 내용](https://developer.apple.com/documentation/uikit/views_and_controls/collection_views/updating_collection_views_using_diffable_data_sources#:~:text=Additionally%2C%20this%20strategy,each%20changed%20item.) 에 따르면, item identifier가 struct일때 특정 셀의 내용을 업데이트하려면 새로운 snapshot을 apply해서 변경이 일어난 부분을 삭제하고, 새로운 내용을 삽입해야합니다.

문서에서 item identifier로 struct를 사용하는 경우는 diffable datasource의 item의 내용이 바뀌지 않는 간단한 경우에만 사용하라는 내용이 있습니다.
<img width="775" alt="스크린샷 2022-08-30 오후 6 56 29" src="https://user-images.githubusercontent.com/60725934/187407907-cc7c1712-8027-48d2-9f1f-03bdb86f8da5.png">

yim2627 commented on MovieFinder/Domain/Entity/MovieListItem.swift:
```diff
@@ -0,0 +1,81 @@
+        default:
+            return nil
+        }
+    }
```
실패가능한 이니셜라이저로 만드신 이유가 있나요?

siwonkim0 replied:
rawValue가 enum case중에 없는 경우에는 nil을 반환해주기 위해서입니다.

yim2627 commented on MovieFinder/Domain/Entity/MovieListItem.swift:
```diff
@@ -0,0 +1,81 @@
+        case .thai:
+            return "THAI"
+        default:
+            return "N/A"
```
이쪽은 타지도 않을 것 같네요.

yim2627 commented on MovieFinder/Domain/UseCase/AccountUseCase.swift:
```diff
@@ -0,0 +1,30 @@
+}
+
+final class AccountUseCase: MoviesAccountUseCase {
+    let accountRepository: AccountRepository
```
이니셜라이저로 주입해주는 프로퍼티인데 이런 부분은 전부 외부 타입에 공개되어있네요. 외부에서 쓰지않는 것 같아요.

yim2627 commented on MovieFinder/Domain/UseCase/AuthUseCase.swift:
```diff
@@ -0,0 +1,38 @@
+
+final class AuthUseCase: MoviesAuthUseCase {
+    let authRepository: MovieAuthRepository
+    let accountRepository: MovieAccountRepository
```
똑같은 문제가 있네요.

yim2627 commented on MovieFinder/Domain/UseCase/DefaultMoviesUseCase.swift:
```diff
@@ -0,0 +1,116 @@
+}
+
+final class DefaultMoviesUseCase: MoviesUseCase {
+    let moviesRepository: MoviesRepository
```
이런 부분은 다 공개되어있네요.

yim2627 commented on MovieFinder/Presentation/Authentication/AuthViewController.swift:
```diff
@@ -0,0 +1,78 @@
+    let disposeBag = DisposeBag()
+    let viewModel: AuthViewModel
+    var coordinator: AuthViewControllerDelegate?
+    let sceneWillEnterForegroundSubject = PublishSubject<Void>()
```
`coordinator`빼고 전부 외부에서 사용하지 않네요.

yim2627 commented on MovieFinder/Presentation/Authentication/AuthViewController.swift:
```diff
@@ -0,0 +1,78 @@
+        NotificationCenter.default.addObserver(self, selector: #selector(AuthViewController.sceneWillEnterForeground), name: UIScene.didActivateNotification, object: nil)
+    }
+    
+    @objc func sceneWillEnterForeground(notification: NSNotification) {
+        sceneWillEnterForegroundSubject.onNext(())
+    }
```
외부에서 사용하지 않네요. `sceneWillEnterForegroundSubject`는 어떤 용도인가요?

siwonkim0 replied:
사용자가 외부 링크를 타고 로그인을 완료하고 나서 앱으로 돌아올 때 
로그인을 완료했다면 자동으로 메인화면으로 전환하기 위해서
AuthVC가 UIScene.willEnterForegroundNotification을 받아 sceneWillEnterForegroundSubject에 이벤트를 전달합니다.

yim2627 commented on MovieFinder/Presentation/Authentication/AuthViewController.swift:
```diff
@@ -0,0 +1,78 @@
+        sceneWillEnterForegroundSubject.onNext(())
+    }
+    
+    func configureBind() {
```
외부에서 사용하지 않아요.

yim2627 commented on MovieFinder/Presentation/Authentication/AuthViewController.swift:
```diff
@@ -0,0 +1,78 @@
+            .observe(on: MainScheduler.instance)
+            .subscribe(with: self, onNext: { (self, _) in
+                self.coordinator?.showTabBarController(at: self)
+            }).disposed(by: disposeBag)
```
보통의 `subscribe`랑 어떤 점이 다른건가요?

siwonkim0 replied:
보통의 `subscribe`는 클로저 내부에서 self를 사용할때 약한 참조 처리를 따로 해줘야하지만, `subscribe(with:)`는 `.withUnretained` 와 동일하게 self를 약하게 참조하고, onNext, onError, onCompleted 클로저 내부에서 self를 옵셔널 바인딩을 하여 옵셔널이 아닌 타입으로 사용할 수 있도록 바꿔주기 때문에 편리하다고 생각하여 사용하였습니다.

yim2627 commented on MovieFinder/Presentation/Authentication/AuthViewController.swift:
```diff
@@ -0,0 +1,78 @@
+            }).disposed(by: disposeBag)
+
+        output.didSaveSessionId
+            .observe(on: MainScheduler.instance)
```
`observe(on:)`과 `subscribe(on:)`의 차이는 무엇인가요? 해당 동작은 왜 메인스레드에서 돌리려 하는건가요?

siwonkim0 replied:
화면전환 로직이라 UI관련된 작업이기 때문에 메인스레드로 바꿔주었습니다~

yim2627 commented on MovieFinder/Presentation/Authentication/AuthViewModel.swift:
```diff
@@ -0,0 +1,62 @@
+            return "invalid session"
+        }
+    }
+}
```
`LocalizedError`를 사용하지 않은 이유가 있나요? 파일을 분리해도 좋을 것 같아요.

yim2627 commented on MovieFinder/Presentation/Authentication/AuthViewModel.swift:
```diff
@@ -0,0 +1,62 @@
+        let didSaveSessionId: Observable<Data>
+    }
+    
+    let useCase: MoviesAuthUseCase
```
외부에서 사용하지 않네요.

yim2627 commented on MovieFinder/Presentation/Authentication/AuthViewModel.swift:
```diff
@@ -0,0 +1,62 @@
+            }
+            .flatMap {
+                self.useCase.saveAccountId()
+            }
```
문제가 있어보이네요. RxSwift를 사용할 때 항상 주의해야할 점들이 무엇이 있을까요?

siwonkim0 replied:
클로저를 사용할때 self를 강하게 참조하여 AuthViewModel이 메모리에서 해제되지 않을 수 있는 문제가 생길 것 같습니다. 따라서 self 약한 참조 처리해주었습니다

yim2627 commented on MovieFinder/Presentation/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+        self.window = window
+    }
+    
+    @discardableResult
+    func start() -> UINavigationController {
+        KeychainManager.shared.checkExistingSession()
+        if KeychainManager.shared.isSessionIdExisting {
+            let tab = tabBarController()
+            window?.rootViewController = tab
+        } else {
+            let auth = authNavigationController()
+            window?.rootViewController = auth
+        }
+        window?.makeKeyAndVisible()
+        return UINavigationController()
+    }
```
보통의 coordinator의 `start()`는 반환값이 없는 것으로 알고있는데, 반환값이 있는 이유가 무엇인가요? `@discardableResult`가 있는 것을 보면 반환값이 사용되는 곳은 한정되어있을 것 같아서요.

yim2627 commented on MovieFinder/Presentation/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+class AppCoordinator: Coordinator, AuthCoordinatorDelegate, MovieListCoordinatorDelegate, SearchCoordinatorDelegate, MyAccountCoordinatorDelegate {
+    var childCoordinators = [Coordinator]()
+    var isloggedIn: Bool = false
+    var window: UIWindow?
```
외부에서 사용하지 않네요.

yim2627 commented on MovieFinder/Presentation/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+protocol Coordinator: AnyObject {
+    var childCoordinators: [Coordinator] { get set }
+    
+    func start() -> UINavigationController
```
모든 `Coordinator`가 `UINavigationController`를 반환받아 사용하나요?

yim2627 commented on MovieFinder/Presentation/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+            vc.navigationController?.pushViewController(detailViewController, animated: true)
+        } else if let vc = viewController as? MyAccountViewController {
+            vc.navigationController?.pushViewController(detailViewController, animated: true)
+        }
```
조건문 내부 동작들이 모두 같은데, 왜 분기를 한건가요?

yim2627 commented on MovieFinder/Presentation/Coordinator/AppCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+    func start() -> UINavigationController
+}
+
+class AppCoordinator: Coordinator, AuthCoordinatorDelegate, MovieListCoordinatorDelegate, SearchCoordinatorDelegate, MyAccountCoordinatorDelegate {
```
이 클래스를 어디선가 상속하나요? 
Coordinator 패턴을 사용할 때 주의해야할 점은 무엇이 있을까요?

yim2627 replied:
Delegate에 따라 extension으로 분리하면 가독성이 더 좋아질 것 같아요.

yim2627 commented on MovieFinder/Presentation/Coordinator/AuthCoordinator.swift:
```diff
@@ -0,0 +1,63 @@
+
+    init() {
+        self.navigationController = UINavigationController()
+    }
```
이니셜라이저를 왜 만들어주신건가요?

yim2627 commented on MovieFinder/Presentation/Coordinator/AuthCoordinator.swift:
```diff
@@ -0,0 +1,63 @@
+class AuthCoordinator: Coordinator, AuthViewControllerDelegate {
+    weak var parentCoordinator: AuthCoordinatorDelegate?
+    var childCoordinators = [Coordinator]()
+    var navigationController: UINavigationController
```
외부에서 사용하지 않네요.

yim2627 commented on MovieFinder/Presentation/Coordinator/MovieListCoordinator.swift:
```diff
@@ -0,0 +1,54 @@
+
+    init() {
+        self.navigationController = UINavigationController()
+    }
```
`AuthCoordinator`랑 같은 궁금증이 드네요.

siwonkim0 replied:
삭제했습니다

yim2627 commented on MovieFinder/Presentation/Coordinator/MyAccountCoordinator.swift:
```diff
@@ -0,0 +1,40 @@
+
+import UIKit
+
+protocol MyAccountCoordinatorDelegate: AnyObject {
+    
+}
```
이 프로토콜은 왜 만들어주신건가요?

siwonkim0 replied:
나중에 MyAccountViewController에서 DetailViewController로 화면전환할때 화면전환 로직을 parentCoordinator에게 맡기기 위해 미리 프로토콜을 만들어놨습니다

yim2627 commented on MovieFinder/Presentation/Coordinator/MyAccountCoordinator.swift:
```diff
@@ -0,0 +1,40 @@
+
+    init() {
+        self.navigationController = UINavigationController()
+    }
```
다른 코디네이터들과 같은 문제가 있네요.

yim2627 commented on MovieFinder/Presentation/Coordinator/SearchCoordinator.swift:
```diff
@@ -0,0 +1,48 @@
+
+    init() {
+        self.navigationController = UINavigationController()
+    }
```
다른 코디네이터들과 같은 문제가 있네요.

yim2627 commented on MovieFinder/Presentation/Extension/ObservableType+extension.swift:
```diff
@@ -0,0 +1,17 @@
+import RxSwift
+
+extension ObservableType {
+    func filterErrors() -> Observable<Element> {
+        return materialize()
+            .filter { $0.element != nil }
+            .dematerialize()
+    }
```
왜 해당 메서드를 만드신건가요? 이 메서드는 어떻게 동작하나요?

yim2627 commented on MovieFinder/Presentation/Extension/UIImage+extension.swift:
```diff
@@ -0,0 +1,45 @@
+            self.draw(in: CGRect(origin: .zero, size: newSize))
+        }
+        
+        print("origin: \(self), resize: \(renderImage)")
```
필요 없는 코드가 있는 것 같네요.

yim2627 commented on MovieFinder/Presentation/Extension/UIImage+extension.swift:
```diff
@@ -0,0 +1,45 @@
+import UIKit
+
+extension UIImage {
+    func resize(newWidth: CGFloat) -> UIImage {
+        let scale = newWidth / self.size.width
+        let newHeight = self.size.height * scale
+
+        let newSize = CGSize(width: newWidth, height: newHeight)
+        let render = UIGraphicsImageRenderer(size: newSize)
+        let renderImage = render.image { context in
+            self.draw(in: CGRect(origin: .zero, size: newSize))
+        }
+        
+        print("origin: \(self), resize: \(renderImage)")
+        return renderImage
+    }
```
해당 메서드가 어떻게 동작하여 이미지를 리사이징하는건가요?

yim2627 commented on MovieFinder/Presentation/Extension/UIImage+extension.swift:
```diff
@@ -0,0 +1,45 @@
+        return renderImage
+    }
+    
+    func downSample(at url: URL, to pointSize: CGSize, scale: CGFloat) -> UIImage {
+        let imageSourceOptions = [kCGImageSourceShouldCache: false] as CFDictionary
+        guard let imageSource = CGImageSourceCreateWithURL(url as NSURL, imageSourceOptions) else {
+            return UIImage()
+        }
+        
+        let maxDimensionInPixels = max(pointSize.width, pointSize.height) * scale
+        let downsampleOptions = [
+            kCGImageSourceCreateThumbnailFromImageAlways: true,
+            kCGImageSourceShouldCacheImmediately: true,
+            kCGImageSourceCreateThumbnailWithTransform: true,
+            kCGImageSourceThumbnailMaxPixelSize: maxDimensionInPixels
+        ] as CFDictionary
+        
+        guard let downsampledImage = CGImageSourceCreateThumbnailAtIndex(imageSource, 0, downsampleOptions) else {
+            return UIImage()
+        }
+        return UIImage(cgImage: downsampledImage)
+    }
```
이 메서드는 왜 만드신거고, 어떻게 동작하는 것인가요?

yim2627 commented on MovieFinder/Presentation/MovieDetail/View Model/MovieDetailViewModel.swift:
```diff
@@ -0,0 +1,128 @@
+        selectedReview.showAllContent.toggle()
+        if let index = reviews?.firstIndex(where: { $0.id == reviewID }) {
+            reviews?[index] = selectedReview
+        }
```
이 로직은 어떻게 동작하는건가요?

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailHeaderView.swift:
```diff
@@ -0,0 +1,45 @@
+import SnapKit
+
+final class MovieDetailHeaderView: UICollectionReusableView {
+    lazy var label: UILabel = {
+        let label = UILabel()
+        label.textColor = .black
+        label.font = UIFont(name: "AvenirNext-Bold", size: 20) ?? UIFont.systemFont(ofSize: 20)
+        return label
+    }()
```
다른 커스텀뷰들은 메서드로 UI를 그릴 때 필요한 데이터를 주입해주시는데, 이 타입은 그러시지 않은 이유가 있나요? 통일되지않은 것 같아서요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailReviewsCollectionViewCell.swift:
```diff
@@ -0,0 +1,81 @@
+final class MovieDetailReviewsCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var userNameLabel: UILabel!
+    @IBOutlet weak var reviewLabel: UILabel!
+    @IBOutlet weak var ratingView: CosmosView!
```
외부에서 참조하지 않는 것 같네요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailReviewsCollectionViewCell.swift:
```diff
@@ -0,0 +1,81 @@
+        self.reviewLabel.frame = CGRect(x: 0, y: 0, width: bounds.width, height: 0)
+        self.reviewLabel.sizeToFit()
+        self.reviewLabel.frame.size = reviewLabel.bounds.size
+    }
```
어떤 동작을 위해 해당 메서드를 오버라이드 하신건가요?
이 메서드는 언제 호출될까요?

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailReviewsCollectionViewCell.swift:
```diff
@@ -0,0 +1,81 @@
+            self.reviewLabel.attributedText = allText
+        } else if !cellViewModel.showAllContent && cellViewModel.content.count > 300 {
+            self.reviewLabel.attributedText = previewText
+        }
```
해당 조건들의 의미가 무엇인가요?

yim2627 commented on MovieFinder/Presentation/MovieList/View/MovieListViewController.swift:
```diff
@@ -0,0 +1,176 @@
+    private let disposeBag = DisposeBag()
+    private var movieListDataSource: DataSource!
+    
+    private typealias DataSource = UICollectionViewDiffableDataSource<Section, MovieListCellViewModel>
+    private typealias Snapshot = NSDiffableDataSourceSnapshot<Section, MovieListCellViewModel>
```
`Section`타입을 클래스로 따로 만들어주신 이유가 궁금해요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailViewController.swift:
```diff
@@ -0,0 +1,357 @@
+        posterImageView.layer.cornerRadius = 20
+        posterImageView.kf.indicatorType = .activity
+        return posterImageView
+    }()
```
`MovieDetailViewController`의 뷰들은 전부 외부에 오픈되어있네요. 따로 이유가 있나요?

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailViewController.swift:
```diff
@@ -0,0 +1,357 @@
+        ratingView.settings.emptyColor = .lightGray
+        ratingView.settings.emptyBorderColor = .lightGray
+        ratingView.settings.starSize = 30
+        return ratingView
```
전부 오픈되어있네요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailViewController.swift:
```diff
@@ -0,0 +1,357 @@
+    let contentView = UIView()
+    
+    private let ratingRelay = BehaviorRelay<Double>(value: 0)
+    private lazy var input = MovieDetailViewModel.Input(viewWillAppear: self.rx.viewWillAppear.asObservable(), tapRatingButton: ratingRelay.asObservable())
```
이 input은 왜 인스턴스 프로퍼티로 빠져있는건가요? 다른 곳은 그러지 않은 것 같아서요.

siwonkim0 replied:
통일성있게 메서드 내로 위치 수정했습니다~!

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailViewController.swift:
```diff
@@ -0,0 +1,357 @@
+        output.imageUrlObservable
+            .observe(on: MainScheduler.instance)
+            .subscribe { self.configureImageView(with: $0) }
+            .disposed(by: disposeBag)
```
여기도 참조 문제가 있어보이네요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailViewController.swift:
```diff
@@ -0,0 +1,357 @@
+        output.ratingDriver
+            .map { $0 ? self.ratingView.rating : 0}
+            .drive(ratingView.rx.rating)
+            .disposed(by: disposeBag)
```
문제가 있어보이네요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailViewController.swift:
```diff
@@ -0,0 +1,357 @@
+            alignment: .top)
+    }
+    
+    func configureLayout() {
```
외부에서 레이아웃을 건드나요? 
해당 메서드는 조금 분리할 필요가 있어보여요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/PlotSummaryCollectionViewCell.swift:
```diff
@@ -0,0 +1,25 @@
+import UIKit
+
+final class PlotSummaryCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var plotSummaryTextView: UITextView!
+    var viewModel: MovieDetailBasicInfo!
```
외부에서 사용하지 않네요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/PlotSummaryCollectionViewCell.xib:
```diff
@@ -0,0 +1,52 @@
+                    <textView clipsSubviews="YES" multipleTouchEnabled="YES" contentMode="scaleToFill" scrollEnabled="NO" editable="NO" selectable="NO" translatesAutoresizingMaskIntoConstraints="NO" id="h9L-Qg-6a2">
+                        <rect key="frame" x="0.0" y="0.0" width="331" height="160"/>
+                        <color key="backgroundColor" systemColor="systemBackgroundColor"/>
+                        <string key="text">Lorem ipsum dolor sit er elit lamet, consectetaur cillium adipisicing pecu, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Ex ...
```
필요없는 테스트성 텍스트는 지워도 될 것 같아요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View Model/MovieDetailViewModel.swift:
```diff
@@ -0,0 +1,128 @@
+    }
+    
+    func transform(_ input: Input) -> Output {
+        let reviewsObservable = input.viewWillAppear
+            .withUnretained(self)
+            .flatMap { (self, _) -> Observable<[MovieDetailReview]> in
+                return self.moviesUseCase.getMovieDetailReviews(from: self.movieID)
+                    .map { reviews -> [MovieDetailReview] in
+                        self.reviews = reviews
+                        return reviews
+                    }
+            }
```
문제가 있을 수 도 있을 것 같네요.

yim2627 commented on MovieFinder/Presentation/MovieList/View Model/MovieListViewModel.swift:
```diff
@@ -0,0 +1,60 @@
+        let refresh: Observable<[Section]>
+    }
+    
+    let defaultMoviesUseCase: MoviesUseCase
```
외부에서 사용하나요?

yim2627 commented on MovieFinder/Presentation/MovieList/View Model/Section.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+final class Section: Hashable {
+    var id = UUID()
+    var title: String
+    var movies: [MovieListCellViewModel]
```
변경되는 프로퍼티인가요? 외부에서 사용하지 않는 것도 보이네요.

siwonkim0 replied:
감사합니다

yim2627 commented on MovieFinder/Presentation/MovieList/View/Cell/MovieListCollectionViewCell.swift:
```diff
@@ -0,0 +1,50 @@
+import Kingfisher
+
+final class MovieListCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var posterImageView: UIImageView!
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var originalLanguageLabel: UILabel!
+    @IBOutlet weak var genresLabel: UILabel!
```
외부에 공개되어있네요.

yim2627 commented on MovieFinder/Presentation/MovieList/View/Cell/MovieListCollectionViewCell.swift:
```diff
@@ -0,0 +1,50 @@
+        self.genresLabel.font = UIFont(name: "AvenirNext-Regular", size: 10) ?? UIFont.systemFont(ofSize: 10)
+    }
+    
+    override func layoutSubviews() {
+        self.posterImageView.layer.cornerRadius = 10
+    }
```
빠진게 있지않나요? 정상 동작하던가요?

yim2627 commented on MovieFinder/Presentation/MovieList/View/MovieListHeaderView.swift:
```diff
@@ -0,0 +1,44 @@
+import UIKit
+
+final class MovieListHeaderView: UICollectionReusableView {
+    let label: UILabel = {
+        let label = UILabel()
+        label.textColor = .black
+        label.font = UIFont(name: "AvenirNext-Bold", size: 20) ?? UIFont.systemFont(ofSize: 20)
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
+    }()
```
통일성이 없는 것 같네요. 어떤 것은 `private`을 풀고 직접 접근하고, 어떤 것은 메서드로 접근하는데 기준이 무엇인가요?

yim2627 commented on MovieFinder/Presentation/MovieList/View/MovieListHeaderView.swift:
```diff
@@ -0,0 +1,44 @@
+        initialize()
+    }
+    
+    func initialize() {
+        self.label.text = nil
+    }
+    
+    func setConstraints() {
+        self.label.snp.makeConstraints { make in
+            make.trailing.bottom.equalToSuperview().offset(10)
+            make.top.leading.equalToSuperview().offset(15)
+        }
+    }
```
외부에서 부르나요?

yim2627 commented on MovieFinder/Presentation/MovieList/View/MovieListViewController.swift:
```diff
@@ -0,0 +1,176 @@
+}
+
+final class MovieListViewController: UIViewController {
+    var collectionView: UICollectionView?
```
외부에서 접근하나요?

yim2627 commented on MovieFinder/Presentation/MovieList/View/MovieListViewController.swift:
```diff
@@ -0,0 +1,176 @@
+    
+    private func setView() {
+        self.view.backgroundColor = .white
+        self.edgesForExtendedLayout = []
```
무엇을 하는 코드인가요? 빈 배열이 들어가있네요.

yim2627 commented on MovieFinder/Presentation/MovieDetail/View/MovieDetailViewController.swift:
```diff
@@ -0,0 +1,357 @@
+    }
+    
+    private func configureDataSource() {
+        self.movieDetailDataSource = DataSource(collectionView: self.collectionView) { [weak self] collectionView, indexPath, itemIdentifier in
+            self?.viewModel.updateReviewState(with: itemIdentifier)
+            let cell = collectionView.dequeueReusableCell(withClass: MovieDetailReviewsCollectionViewCell.self, indexPath: indexPath)
+            guard let reviews = self?.viewModel.reviews,
+                  let review = reviews.filter({ $0.id == itemIdentifier }).first else {
+                return MovieDetailReviewsCollectionViewCell()
+            }
+            cell.configure(with: review)
+
+            return cell
+        }
```
`weak self`를 해주신 이유가 있을까요?

yim2627 commented on MovieFinder/Presentation/MovieList/View/SectionBackgroundDecorationView.swift:
```diff
@@ -0,0 +1,18 @@
+
+import UIKit
+
+final class SectionBackgroundDecorationView: UICollectionReusableView {
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
```
무엇을 하는 타입인가요? 아무것도 없네요.

siwonkim0 replied:
사용하지 않아 삭제했습니다

yim2627 commented on MovieFinder/Presentation/MyAccount/MyAccountViewController.swift:
```diff
@@ -0,0 +1,23 @@
+
+import UIKit
+
+protocol MyAccountViewControllerDelegate {
+    
+}
```
아무것도 하지 않는 것처럼 보이는데 용도가 무엇인가요?

yim2627 commented on MovieFinder/Presentation/MyAccount/MyAccountViewController.swift:
```diff
@@ -0,0 +1,23 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+//        viewModel.getAccountDetails(sessionID: "23a5123518359f5a9d2bb715be90c4703cdc2bca")
```
필요없는 코드는 지워주세요.

yim2627 commented on MovieFinder/Presentation/MyAccount/MyAccountViewController.swift:
```diff
@@ -0,0 +1,23 @@
+}
+
+final class MyAccountViewController: UIViewController {
+    let viewModel = MyAccountViewModel()
```
외부에서 접근하나요?

yim2627 commented on MovieFinder/Presentation/Search/View Model/SearchViewModel.swift:
```diff
@@ -0,0 +1,56 @@
+        let searchCancelledObservable: Observable<[SearchCellViewModel]>
+    }
+    
+    let apiManager = URLSessionManager()
+    let useCase: MoviesUseCase
```
외부에서 접근하나요?

yim2627 commented on MovieFinder/Presentation/Search/View/RecentSearchCollectionViewCell.swift:
```diff
@@ -0,0 +1,57 @@
+import SnapKit
+
+final class RecentSearchCollectionViewCell: UICollectionViewCell {
+    let posterImageView: UIImageView = {
+        let poster = UIImageView()
+        poster.layer.cornerRadius = 5
+        return poster
+    }()
+    
+    let titleLabel: UILabel = {
+        let title = UILabel()
+        title.text = "title"
+        return title
+    }()
+    
+    let genresLabel: UILabel = {
+        let genres = UILabel()
+        genres.text = "genres"
+        return genres
+    }()
+    
+    let overviewLabel: UILabel = {
+        let overview = UILabel()
+        overview.text = "overview"
+        return overview
+    }()
```
외부에서 사용하나요?

yim2627 commented on MovieFinder/Presentation/Search/View/RecentSearchCollectionViewCell.swift:
```diff
@@ -0,0 +1,57 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    func setLayout() {
+        let labelStackView = UIStackView(arrangedSubviews: [titleLabel, genresLabel, overviewLabel])
+        labelStackView.axis = .vertical
+        posterImageView.setContentHuggingPriority(.init(rawValue: 255), for: .horizontal)
+        let entireStackView = UIStackView(arrangedSubviews: [posterImageView, labelStackView])
+        entireStackView.distribution = .fillEqually
+        contentView.addSubview(entireStackView)
+        
+        entireStackView.snp.makeConstraints { make in
+            make.edges.equalToSuperview()
+        }
+    }
```
외부에서 사용하나요?

yim2627 commented on MovieFinder/Presentation/Search/View/SearchCollectionViewCell.swift:
```diff
@@ -0,0 +1,39 @@
+import UIKit
+import Kingfisher
+
+class SearchCollectionViewCell: UICollectionViewCell {
```
상속하나요?

yim2627 commented on MovieFinder/Presentation/Search/View/SearchCollectionViewCell.swift:
```diff
@@ -0,0 +1,39 @@
+import Kingfisher
+
+class SearchCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var posterImageView: UIImageView!
+    @IBOutlet weak var titleLabel: UILabel!
+    @IBOutlet weak var decriptionLabel: UILabel!
```
외부에서 사용하나요?

yim2627 commented on MovieFinder/Presentation/Search/View/SearchTableViewController.swift:
```diff
@@ -0,0 +1,48 @@
+import RxSwift
+import RxCocoa
+import SnapKit
+
+final class SearchTableViewController: UIViewController, UITableViewDataSource, UISearchResultsUpdating {
+    let tableView = UITableView()
```
외부에서 사용하지 않네요.

yim2627 commented on MovieFinder/Presentation/Search/View/SearchTableViewController.swift:
```diff
@@ -0,0 +1,48 @@
+        tableView.dataSource = self
+    }
+    
+    func setupConstraints() {
+        tableView.snp.makeConstraints { make in
+            make.leading.trailing.top.bottom.equalToSuperview()
+        }
+    }
```
외부에서 사용하지 않네요.

yim2627 commented on MovieFinder/Presentation/Search/View/SearchTableViewController.swift:
```diff
@@ -0,0 +1,48 @@
+    }
+    
+    func updateSearchResults(for searchController: UISearchController) {
+        let searchText = searchController.searchBar.text!
```
아무것도 하지않는 코드 같은데 용도가 무엇인가요?

yim2627 commented on MovieFinder/Presentation/Search/View/SearchViewController.swift:
```diff
@@ -0,0 +1,144 @@
+    
+    private typealias DataSource = UICollectionViewDiffableDataSource<Section, SearchCellViewModel>
+    private typealias Snapshot = NSDiffableDataSourceSnapshot<Section, SearchCellViewModel>
+    private lazy var input = SearchViewModel.Input(
+        viewWillAppear: self.rx.viewWillAppear.asObservable(),
+        searchBarText: searchController.searchBar.rx.text.orEmpty.asObservable(),
+        searchCancelled: searchController.searchBar.rx.cancelButtonClicked.asObservable()
+    )
```
인스턴스 프로퍼티로 두신 이유가 무엇인가요?

yim2627 commented on MovieFinder/Presentation/Search/View/SearchViewController.swift:
```diff
@@ -0,0 +1,144 @@
+    private let viewModel: SearchViewModel
+    private let disposeBag = DisposeBag()
+    private var searchDataSource: DataSource!
+    let cancelButtonClicked = PublishSubject<Void>()
```
외부에서 접근하지않네요!!!!!!!!!!!!!!!!

yim2627 commented on MovieFinder/Presentation/Search/View/SearchViewController.swift:
```diff
@@ -0,0 +1,144 @@
+        searchCancelled: searchController.searchBar.rx.cancelButtonClicked.asObservable()
+    )
+    
+    lazy var collectionView: UICollectionView = {
+        let layout: UICollectionViewFlowLayout = UICollectionViewFlowLayout()
+        layout.scrollDirection = .vertical
+        layout.itemSize = CGSize(width: view.frame.width - CGFloat(40), height: 110)
+        layout.minimumInteritemSpacing = 0.0
+        layout.minimumLineSpacing = 1.0
+        
+        let collectionView = UICollectionView(frame: view.frame, collectionViewLayout: layout)
+        return collectionView
+    }()
+    
+    lazy var searchController: UISearchController = {
+        let resultController = SearchTableViewController()
+        let searchController = UISearchController(searchResultsController: nil)
+        searchController.searchResultsUpdater = self
+        searchController.obscuresBackgroundDuringPresentation = false
+        searchController.searchBar.placeholder = "Search Movies"
+        definesPresentationContext = true
+        return searchController
+    }()
```
private

yim2627 replied:
밑에도 알아서 붙이세요

yim2627 commented on MovieFinder/Util/DownloadableUIImageView.swift:
```diff
@@ -0,0 +1,50 @@
+    
+    func getImage(with url: URL?) {
+        self.image = UIImage()
+        self.layer.cornerRadius = CGFloat(15)
```
이 메서드에 어울리지 않는다고 생각되네요.

yim2627 commented on MovieFinder/Data/Network/URLSessionManager.swift:
```diff
@@ -0,0 +1,50 @@
+        self.urlSession = urlSession
+    }
+    // MARK: - Networking
+    func performDataTask<T: NetworkRequest>(with requestType: T) -> Observable<T.ResponseType> {
+        return Observable.create { observer in
+            guard let request = requestType.urlRequest else {
+                observer.onError(URLSessionError.invalidRequest)
+                return Disposables.create()
+            }
+            let task = self.urlSession.dataTask(with: request) { data, response, error in
+                if let error = error {
+                    observer.onError(URLSessionError.requestFailed(description: error.localizedDescription))
+                    return
+                }
+                if let httpResponse = response as? HTTPURLResponse,
+                   httpResponse.statusCode >= 300 {
+                    observer.onError(URLSessionError.responseFailed(code: httpResponse.statusCode))
+                    return
+                }
+                guard let data = data else {
+                    observer.onError(URLSessionError.invaildData)
+                    return
+                }
+                guard let decodedData = JSONParser.decodeData(of: data, type: T.ResponseType.self) else {
+                    return
+                }
+                observer.onNext(decodedData)
+            }
+            task.resume()
+            return Disposables.create {
+                task.cancel()
+            }
+        }
+    }
```
문제가 좀 있을 것 같네요.

yim2627 commented on MovieFinder/Data/Repository/AccountRepository.swift:
```diff
@@ -0,0 +1,112 @@
+        self.urlSessionManager = urlSessionManager
+    }
+    
+    func saveAccountId() -> Observable<Data> {
+        let accountIdRequest = AccountIdRequest(
+            queryParameters: [
+                "api_key": ApiKey.tmdb.description,
+                "session_id": KeychainManager.shared.getSessionID()
+            ]
+        )
+        return urlSessionManager.performDataTask(with: accountIdRequest)
+            .map { $0.id }
+            .map { accountID in
+                guard let dataSessionID = String(accountID).data(
+                        using: String.Encoding.utf8,
+                        allowLossyConversion: false) else {
+                    return Data()
+                }
+                self.saveAccountIDToKeyChain(dataSessionID)
+                return dataSessionID
+            }
+    }
```
문제가.. 있겠죠

yim2627 commented on MovieFinder/Data/Repository/AccountRepository.swift:
```diff
@@ -0,0 +1,112 @@
+            }
+    }
+    
+    private func getTotalRatedList() -> Observable<[MovieListItemDTO]> {
+        return getTotalRatedListPages()
+            .flatMap { totalPages in
+                Observable.zip((1...totalPages).map { page in
+                    self.getRatedMovieList(page: page) }
+                )
+            }
+            .map { arrays in
+                arrays.flatMap { $0 }
+            }
+    }
```
어떤 문제가 있을까요?

yim2627 commented on MovieFinder/Data/Repository/DefaultMoviesRepository.swift:
```diff
@@ -0,0 +1,70 @@
+                "api_key": ApiKey.tmdb.description
+            ]
+        )
+        return self.urlSessionManager.performDataTask(with: reviewsRequest)
```
self를 붙일 때의 기준을 명확히 하는게 좋아보여요.

yim2627 reviewed (commented):
최대한 자세히 보려고 봤지만 빠진 부분이 있을 수 있으니 이해부탁드려요.

로직 관련 피드백은 제가 정확한 프로젝트 구조를 알지 못하여서 모르는 부분은 질문드렸고, 다시 되새기면서 정리한다는 느낌으로 가주심 좋을 것 같아요.

일단 가장 큰 문제는 모델 문제, 메모리릭 정도가 있는 것 같아요.

## 모델 디코딩 문제
![image](https://user-images.githubusercontent.com/70251136/187026755-e0be42e3-7a4b-4dfe-88c6-9dc91c514e15.png)

이전에 말씀드렸던 API 타입문제입니다.

## 메모리 릭
<img width="709" alt="스크린샷 2022-08-27 오후 7 46 54" src="https://user-images.githubusercontent.com/70251136/187026856-52533d44-fe0d-470b-a7bf-f8acbc8082a9.png">

siwonkim0 committed:
feat: 검색 결과 pagination 기능 추가 #15

- collectionView의 contentOffset 이벤트를 구독하여 스크롤의 위치가 일정 부분을 넘었을 때 다음 페이지 검색결과를 받아오도록 구현함
- 스크롤을 내린 후에 cancel 버튼을 누르면 이전 검색 결과가 표시되는 버그 있음
- 스크롤을 내리면 몇개의 페이지가 한꺼번에 표시되거나 중복된 결과가 들어오는 버그 있음

siwonkim0 committed:
refactor: 필요없는 분기처리 제거

siwonkim0 committed:
refactor: MovieDetailViewController 순환참조 문제 해결

- 몇개의 클로저에서 vc를 강하게 참조하고 있어서 네비게이션 컨트롤러에서 pop되어도 메모리에는 남아있는 memory leak 문제가 발생하여서 약한 참조로 변경해주었음

siwonkim0 committed:
del: 사용하지 않는 sectionBackground 삭제

siwonkim0 committed:
refactor: model의 프로퍼티 변경

- 기존 MovieListItem을 들고있는 MovieList 타입을 새로 만들어서 page와 totalPages 정보를 가지고 있도록 함

siwonkim0 committed:
refactor: diffable datasource 중복된 item이 들어가서 생기는 오류 수정 #15

siwonkim0 committed:
refactor: json 데이터 디코딩 모델 타입 프로퍼티 옵셔널로 변경

- API 명세에 따라 옵셔널로 변경하여 값이 없는 경우에도 디코딩이 잘 되도록 변경

siwonkim0 committed:
refactor: searchBar의 cancel 버튼을 누르면 현재 검색어로 pagination이 되는 오류 해결 #15

- viewModel에 canLoadNextPage라는 flag를 두고, searchBar.text 이벤트가 들어오면 canLoadNextPage = true, cancel버튼 이벤트가 들어오면 canLoadNextPage = false로 두어서 pagination 이벤트가 불릴때는 오직 canLoadNextPage = true인 상태에서만 실행이 되도록 수정함
- viewModel에 searchResults behaviorRelay를 두어서 SearchBar.text, cancelButton, pagination 이벤트가 각각 들어오면 api 통신 결과를 searchResults에 onNext로 전달함
- viewController는 searchResults만 구독하도록 수정함

siwonkim0 committed:
refactor: AuthVC가 메모리에서 해제되지 않는 문제 해결

- AuthVC가 AuthCoordinator을 강하게 참조하고 있어서 AppCoordinator의 child에서 AuthCoordinator을 해제해도 AuthCoordinator가 메모리에서 해제되지 않아 AuthVC, VM, useCase, repo 모두 해제되지 않는 상황 발생
- 약한 참조로 변경하여 해결

siwonkim0 committed:
chore: 클로저 내부를 제외하고 self 제거하여 컨벤션 통일함

siwonkim0 committed:
del: 사용하지 않는 메서드 삭제

siwonkim0 committed:
refactor: 인코딩하지 않는데 Codable을 채택해 준 모델에 Decodable을 채택하도록 변경함

siwonkim0 committed:
refactor: 옵셔널 값을 반환하던 부분을 함수 내부에서 처리하도록 변경함

siwonkim0 committed:
del: enum의 연관값 파라미터 삭제

siwonkim0 committed:
chore: 파일 주석에 파일명을 타입명과 동일하게 수정함

siwonkim0 committed:
chore: 외부에서 접근하지 않는 프로퍼티에 private 키워드 추가

siwonkim0 committed:
del: 필요없는 이니셜라이저 삭제

siwonkim0 committed:
refactor: 로그인하지 않고 AuthVC로 돌아오는 경우 error가 발생해 stream이 끊어지는 문제 해결 #14

- retry() 사용하여 error가 발생해도 다시 시도하도록 변

siwonkim0 committed:
del: 사용하지 않는 코드 삭제

siwonkim0 committed:
refactor: DetailViewModel의 Observable -> Driver로 변경

- UI업데이트를 위해 사용하는 Observable을 share과 main thread에서 작동하는 기능을 가진 driver로 변경

siwonkim0 committed:
del: 사용하지 않는 코드 삭제

siwonkim0 committed:
chore: 주석 삭제

siwonkim0 committed:
refactor: cancel 버튼을 누르면 pagination 이벤트가 불리는 현상 근본적으로 해결

- 기존에는 canLoadNextPage 라는 flag를 두어서 cancel 버튼이 눌리면 canLoadNextPage = false로 두고, pagination을 할때 canLoadNextPage == true 여야 api 호출을 하도록 했는데, 근본적인 원인이 cancel 버튼을 누르면 collectionView의 contentview height가 0이 되어 self.collectionView.frame.height + offset.y + 500 >= self.collectionView.contentSize.height 이 성립되어 pagination 이벤트가 트리거되는 것이었다.
- 따라서 guard문으로 contentsize height가 0인 경우에는 return 하도록하여 이벤트가 발생하지 않도록 하여 해결하였다

siwonkim0 committed:
del: Hashable일 필요없는 모델에 Hashable 삭제

siwonkim0 committed:
test: DefaultMoviesUseCase 테스트 구현

- MockDefaultMoviesRepository 생성하여 DefaultMoviesUseCase에 주입하여 테스트 진행

siwonkim0 committed:
test: AccountUseCase 테스트 구현

- MockAccountRepository 생성하여 AccountUseCase에 주입하여 테스트 진행

siwonkim0 committed:
test: AuthUseCase 테스트 구현

- MockAuthRepository 생성하여 AuthUseCase에 주입하여 테스트 진행

siwonkim0 committed:
test: MovieListViewModel 테스트 구현

- MockDefaultMoviesUseCase 생성

siwonkim0 committed:
test: MovieDetailViewModel 테스트 구현

- MockAccountUseCase 생성

siwonkim0 committed:
test: AuthViewModel 테스트 구현

- MockAuthUseCase 생성

siwonkim0 committed:
chore: 접근제어 수정

- 외부에서 쓰지 않는 프로퍼티 private 접근제어자 설정함

siwonkim0 committed:
chore: 접근제어 수정

- 외부에서 쓰지 않는 프로퍼티 private 접근제어자 설정함

siwonkim0 committed:
chore: private, weak self 추가

siwonkim0 committed:
chore: private, weak self 추가

siwonkim0 committed:
refactor: 서치바에 텍스트가 입력되고 0.5초 후에 이벤트가 발생하도록 수정

- 처음 이벤트가 들어오고 0.5초 후에 발생하도록 수정하여 불필요한 api호출을 방지

siwonkim0 committed:
Update README.md

siwonkim0 committed:
feat: Kingfisher disk storage limit 지정함 #6

- downsampling한 이미지는 memory에 저장되고 원본 이미지는 disk에 저장되도록 설정하였기 때문에 disk 저장이 1기가가
넘지 않도록 limit를 걸어주었다.
- expiration date는 기본값이 7일이다.

siwonkim0 committed:
chore: 클로저 외부에서 self 참조시 self 키워드 생략

siwonkim0 committed:
chore: 오타 수정

siwonkim0 committed:
feat: custom cell prepareForReuse에 cancelDownloadTask 추가 #6

siwonkim0 committed:
Update README.md

siwonkim0 committed:
refactor: collectionView tap 이벤트를 input으로 전달하여 viewModel에서 처리하도록 수정

- 코드 일관성을 위해서 기존에는 collectionView tap 이벤트를 viewController에서 바로 subscribe 해주었던 부분을 input으로 넘겨서 viewModel이 모델의 상태를 변경하고, output으로 전달하면 viewController가 화면을 업데이트하도록 수정함

siwonkim0 committed:
fix: DetailView의 collectionView cell configure 시점에 preview를 보여주도록 수정 #12

- previewText를 만드는 부분에 오타로 인해 content 전체가 들어갔던 부분을 contentPreview가 들어가도록 수정함

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
refactor: viewModel에 behaviorRelay 제거후 input을 output으로 변환하여 반환 #15

- input을 받아서 api 호출 결과를 behaviorRelay에 accept해서 viewcontroller로 보내던 부분을 input을 output으로 변형해서 내보내도록 변경함

siwonkim0 committed:
design: SearchViewController customcell에 커스텀 폰트 적용 #15

- imageView cornerRadius 도 변경함

siwonkim0 committed:
feat: Account View 평점목록 구현 #13

siwonkim0 committed:
feat: Account View 평점목록 터치시 상세화면 이동 구현 #13

siwonkim0 committed:
feat: Account View에서 평점 등록 수정하기 구현 #13

- 평점 뷰를 터치하면 평점도 등록되고 상세화면으로 전환도 같이 되는 현상 발생
- 화면 전환을 포스터 이미지 터치시 시키도록 수정 고려중

siwonkim0 committed:
design: 사용하지 않는 버튼 뷰에서 제거

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
refactor: caseIterable을 채택한 enum의 .allCases 사용하여 case 불러오는 것으로 수정 #8

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
refactor: Driver -> Signal로 변경 #13

- Account VC에서 평점뷰를 터치하면 VM이 평점을 서버에 업데이트하고 완료 이벤트를 VC로 전달할때 상태값이 아니라 완료 이벤트를 전달하는 것이므로 구독했을때 최신 이벤트를 replay하는 driver보다는 signal이 적절하다고 판단함
- test: 상태검증, 행위검증을 둘다 하는 객체는 mock이 아니라 spy로 네이밍하는것이 더 적절하다고 판단하여 네이밍 변경함

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
refactor: Observable -> Driver로 변경 #12

- UI 업데이트하기 때문에 driver로 변경함

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
chore: 오타 수정

siwonkim0 committed:
chore: 폴더 구조 변경

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md

siwonkim0 committed:
Update README.md
